test/e2e/Test/E2ESupport.hs

```diff
@@ -154,5 +149,35 @@ runTestCommand' bin args inputLines = do
                 <> err
                 <> "-------------------\n"
 
+-- | Like 'runTestCommandForError'' but does not provide input to stdin.
+runTestCommandForError :: (HasCallStack) => FilePath -> [Text] -> TestM Text
+runTestCommandForError bin args = runTestCommandForError' bin args []
+
+-- | Like 'runTestCommand'' but expects the command to return an error.
+--
+-- If the command exits with a non-zero exit code, an exception is thrown.
+-- Otherwise stdout of the error is returned.
+runTestCommandForError' :: (HasCallStack) => FilePath -> [Text] -> [Text] -> TestM Text
+runTestCommandForError' bin args inputLines = do
+    let argsString = map toS args
+    (exitCode, out, _) <- executeCommand bin argsString inputLines
+    case exitCode of
+        ExitFailure _ -> pure $ toS $ trimNewline out
+        ExitSuccess ->
+            let commandLine = bin <> " " <> unwords argsString
+            in liftIO $ HUnit.assertFailure $
+                "Command " <> commandLine <> " succeeded,\n"
+                <> "but was expected to fail."
+
+executeCommand :: FilePath -> [String] -> [Text] -> TestM (ExitCode, String, String)
```

Comment:
Can you change the signature so that `argsString :: [Text]` to remove the duplicate `let argsString = map toS args` at the call sites.